### ISSUE TITLE
ioctl(dev, DIOCCHANGERULE, ...) PF_CHANGE_ADD_TAIL: File exists

### ISSUE THREAD

**Issue by nakal** [miniupnp/miniupnp#715]
System: FreeBSD 14.0

My Playstation reported NAT Type 3, so I know something is broken with UPNP. I looked in the logs and I see that something is logged right after the connection test.

It was working once. I'm using this anchor rule in `/etc/pf.conf`:

```
rdr-anchor "miniupnpd"
```

ALTQ is disabled.

Log when NAT test on PlayStation 5 runs:

```
Mar 10 09:59:23 hostname miniupnpd[56714]: ioctl(dev, DIOCCHANGERULE, ...) PF_CHANGE_ADD_TAIL: File exists
Mar 10 09:59:25 hostname syslogd: last message repeated 3 times
```

**Comment by miniupnp** [comment]
I'm still using FreeBSD 13.3 ;)

I guess I have to work on the libpf usage quick...

**Comment by miniupnp** [comment]
@nakal just to make sure : have you rebuilt miniupnpd after each FreeBSD major upgrade ?

**Comment by nakal** [comment]
Yes. I build packages from Poudriere. I also tried `pkg install -f miniupnpd`.

Here is the timestamp of the package repo from where I install:

```
-rw-r--r--  1 root wheel 78990 Feb 29 21:24 miniupnpd-2.3.3_3,1.pkg
```

**Comment by dramirezRT** [comment]
I just upgraded to 24.03-RELEASE of pfsense and I am seeing this same behavior.
ioctl(dev, DIOCCHANGERULE, ...) PF_CHANGE_ADD_TAIL: File exists 

Also tried reinstalling the miniupnpd package.

Any idea what the issue could be?

**Comment by miniupnp** [comment]
I guess miniupnpd is still not compatible with FreeBSD 15.
development is still in progress.

**Comment by DarkArc** [comment]
This seems to be affecting pfSense as they've moved to FreeBSD 15.

**Comment by miniupnp** [comment]
@dramirezRT could you try the newest miniupnpd sources from the `master`  branch ?

**Comment by miniupnp** [comment]
it is included in miniupnpd 2.3.7.

**Comment by nakal** [comment]
Sorry for the late feedback. Playstation shows NAT type 2 again. Thank you for the patch.

I've just submitted a problem report to update the package.

**Comment by nakal** [comment]
I accidentally opened the bug.

**Comment by kgcgs** [comment]
I'm seeing the same issue on FreeBSD 14.3 (and 2) running 2.3.9.

# miniupnpd --version
miniupnpd 2.3.9 Jun  9 2025
using pf backend
build options: NAT-PMP PCP check_portinuse

Jun  9 18:14:35 a miniupnpd[1966]: it is advised to use network interface name instead of 10.0.3.1/24
Jun  9 18:14:35 a miniupnpd[1967]: HTTP listening on port 11426
Jun  9 18:14:35 a miniupnpd[1967]: Listening for NAT-PMP/PCP traffic on port 5351
Jun  9 18:26:57 a miniupnpd[1967]: could not find nat rule to delete iport=41641 addr= X
Jun  9 18:26:57 a miniupnpd[1967]: ioctl(dev, DIOCCHANGERULE, ...) PF_CHANGE_ADD_TAIL: File exists
Jun  9 20:26:58 a miniupnpd[1967]: remove port mapping 41642 UDP because it has expired
Jun  9 20:26:58 a miniupnpd[1967]: could not find nat rule to delete iport=41641 addr= X

#NAT PNP For Tailscale
rdr-anchor miniupnpd
anchor miniupnpd

**Comment by kgcgs** [comment]
Yes, that looks to be the default when build from the ports tree.

 * FreeBSD meno 14.2-RELEASE-p1 FreeBSD 14.2-RELEASE-p1 GENERIC amd64
 * using command line options --firewall=pf --libpfctl --portinuse */

```
# ldd /bin/miniupnpd
/bin/miniupnpd:
        libpfctl.so.0 => /usr/local/lib/libpfctl.so.0 (0x3022fd21a000)
        libkvm.so.7 => /lib/libkvm.so.7 (0x3022fc2ab000)
        libc.so.7 => /lib/libc.so.7 (0x3022fe043000)
        libnv.so.1 => /lib/libnv.so.1 (0x3022fe728000)
        libelf.so.2 => /lib/libelf.so.2 (0x3022ff37c000)
        [vdso] (0x3022fbcdc000)
```

**Comment by miniupnp** [comment]
` ioctl(dev, DIOCCHANGERULE, ...) PF_CHANGE_ADD_TAIL: File exists` (`EEXIST`)
is either in `add_filter_rule2()` or `add_nat_rule()`
I guess it is `add_filter_rule2()` should be changed to use `pfctl_add_rule()`

**Comment by kgcgs** [comment]
Let me know if there is anything I can do to help move this forward.

**Comment by kgcgs** [comment]
A bug bounty perhaps?
